# Variant Annotation
Single nucleotide variants and small insertions and deletions were annotated using the Ensembl Variant Effect Predictor (VEP) software. Variant annotation was based on canonical transcripts as defined in Ensembl VEP. Population-level frequencies from Genome Aggregation Database (gnomAD) were added using ANNOVAR. Pathogenicity predictions for stop gain or stop loss, frameshifts and splice site disrupting variants were generated by LOFTEE as part of the VEP pipeline. Pathogenicity predictions for missense variants annotated with PolyPhen-2 webtool. 

Code here is based on GRCh37 (hg19) genomic coordinates.

## Requirements
- PERL
- [Ensembl VEP](https://asia.ensembl.org/info/docs/tools/vep/script/vep_download.html) (v104) and [non-indexed cache](https://asia.ensembl.org/info/docs/tools/vep/script/vep_cache.html) database (v104) for GRCh37
- [dbNSFP](https://sites.google.com/site/jpopgen/dbNSFP) (v3.5a) for VEP plugin
- [LOFTEE](https://github.com/konradjk/loftee) (v1.0.3) for VEP plugin
- [ANNOVAR](https://annovar.openbioinformatics.org/en/latest/) (dated version "2020-06-07" used at time of analysis) to extract Genome Aggregation Database (gnomAD) population-level frequencies (v2.1.1)
- [PolyPhen-2](http://genetics.bwh.harvard.edu/pph2/) tool (v2.2.3r406)


## Usage
### 1. Generate unique identifier for VEP submission from `variants.filtered_DPGQ.vcf.gz`. (Already done in "Steps Prior to Variant Quality Filtering", see step #2)
```bash
./run_uniqueID_vcf.sh
```
Input file: variants.filtered_DPGQ.vcf.gz \
Output file: variants.filtered_DPGQ.uniqID.vcf.gz

<br/>


### 2. Variants annotation was carried out using VEP
```bash
${VEP} -i ${VCF_IN} --cache --dir_cache ${CACHE_DB} --offline --cache_version 104 --use_given_ref \
 --assembly GRCh37 \
 --canonical \
 --numbers \
 --variant_class --hgvs --check_existing \
 --fasta ${REFERENCE} --symbol \
 --plugin dbNSFP,${DBNSFP},${REPLACE_LOGIC},Ensembl_transcriptid,Uniprot_acc_Polyphen2,Polyphen2_HDIV_pred,Polyphen2_HVAR_pred \
 --dir_plugins ${VEP_plugin} --plugin LoF,loftee_path:${LOFTEE},human_ancestor_fa:${LOFTEE_DB}/human_ancestor.fa.gz,conservation_file:${LOFTEE_DB}/phylocsf_gerp.sql,gerp_bases:${LOFTEE_DB}/GERP_scores.final.sorted.txt.gz,gerp_exons:${LOFTEE_DB}/GERP_scores.exons.txt.gz \
 --output_file ${OUT} –tab –verbose
```

Input file: variants.filtered_DPGQ.uniqID.vcf.gz \
Output files: variants.DP8GQ20_filtered.uniqID.vep.table.out*

<br/>


### 3. Filter VEP output & extract relevant columns 
```bash
## to extract required VEP annotation columns
perl VEPannot_extract.pl variants.DP8GQ20_filtered.uniqID.vep.table.out variants.DP8GQ20_filtered.uniqID.vep.table.filtered.out
```

Input: variants.DP8GQ20_filtered.uniqID.vep.table.out \
Output: variants.DP8GQ20_filtered.uniqID.vep.table.filtered.out (See [`variants.DP8GQ20_filtered.uniqID.vep.table.filtered.out.descriptor`](/Variant_annot/variants.DP8GQ20_filtered.uniqID.vep.table.filtered.out.descriptor))

<br/>


### 4. Genome Aggregation Database (gnomAD) population-level frequencies (v2.1.1) were added using ANNOVAR
Format input file `variants.filtered_DPGQ.uniqID.vcf` for ANNOVAR according to https://annovar.openbioinformatics.org/en/latest/user-guide/input/ with modifications

```bash
## we have modified output of convert2annovar.pl, to facilitate merging of annotation output later
## convert2annovar.pl is provided as part of ANNOVAR package
perl ${ANNOVAR_DIR}/convert2annovar.pl -format vcf4 -allsample -withfreq -includeinfo variants.DP8GQ20_filtered.uniqID.vcf > variants.avlist.tmp
cut -f1-8,11 variants.avlist.tmp > variants.avlist

## using ANNOVAR to extract gnomAD exome, gnomAD genome population AF
## to use *_dropped for combine step later
perl ${ANNOVAR_DIR}/annotate_variation.pl -build hg19 -filter -dbtype gnomad211_exome $outdir/variants.avlist -otherinfo $ANNOVAR_DIR/humandb/
perl ${ANNOVAR_DIR}/annotate_variation.pl -build hg19 -filter -dbtype gnomad211_genome $outdir/variants.avlist -otherinfo $ANNOVAR_DIR/humandb/
```

<br/>


### 5. PolyPhen-2 predictions
PolyPhen-2 HDIV predictions for missense variants were [batch queried](http://genetics.bwh.harvard.edu/pph2/bgi.shtml) with the options of: 
- Classified model: HumDiv
- Genome assembly: GRCh37/hg19
- Transcripts: All
- Annotations: All

Predictions based on canonical ENST_ID used. See [`resources/allcanonicalvariants_ID_PolyPhen2.list`](/resources/allcanonicalvariants_ID_PolyPhen2.list). 

<br/>


### 6. Combine variant annotated list with variant genotypes
Combine VEP annotation, population level gnomAD frequencies, PolyPhen-2 prediction together with genotypes of pass QC variants to give the matrix (see example files [`variantannot_samplegeno.tsv`](variantannot_samplegeno.tsv), [`samplegeno.tsv`](samplegeno.tsv)) with the following headers:


#### (a) Combine variant annotation outputs
``` bash
perl merge_VEP_gnomad.pl variants.DP8GQ20_filtered.uniqID.vep.table.filtered.out variants.avlist.hg19_gnomad211_exome_dropped variants.avlist.hg19_gnomad211_genome_dropped
```

Input files: variants.DP8GQ20_filtered.uniqID.vep.table.filtered.out, variants.avlist.hg19_gnomad211_exome_dropped, variants.avlist.hg19_gnomad211_genome_dropped \
Output file: variantannot.tsv



#### (b) Combine variantannot.tsv + samplegenotype_sorted_pmiss_hwe
``` bash
perl merge_annot_geno-pmiss-hwe.pl variantannot.tsv samplegenotype_sorted_pmiss_hwe
```

Input files: variantannot.tsv, samplegenotype_sorted_pmiss_hwe (from Variant_processing step) \
Output files: samplegeno.tsv, variantannot_samplegeno.tsv (See [`variantannot_samplegeno.tsv.descriptor`](/Variant_annot/variantannot_samplegeno.tsv.descriptor))
